web/yo/app/scripts/controllers/gene.js

```diff
@@ -3009,6 +2987,119 @@ angular.module('oncokbApp')
+            return DatabaseConnector.getAllDrugs()
+                .then(function(result){
+                    $scope.addTherapyError = false;
+                    var drugs = result;
+                    return drugs.filter(function (drug) {
+                        return (drug.drugName.toLowerCase().indexOf($query.toLowerCase()) != -1) || (drug.synonyms.join(',').toLowerCase().indexOf($query.toLowerCase()) != -1);
+                    })
+                })
+                .catch(
+                    function (error) {
+                        $scope.therapyErrorMessage = "Sorry, the list of drugs is not available now.";
+                        $scope.addTherapyError = true;
+                    }
+                )
+        }
+        $scope.addDruginTherapy = function(tag,index){
+            $scope.therapy[index].push(tag.drugName);
+            addTherapy(index);
+        }
+
+        $scope.removeDruginTherapy = function(tag,index){
+            var id = $scope.therapy[index].indexOf(tag.drugName);
+            if (id > -1) {
+                $scope.therapy[index].splice(id, 1);
+            }
+            addTherapy(index);
+        }
+        function addTherapy(index) {
+            if ($scope.therapy.length === index + 1 ){
+                $scope.therapy.push([]);
+            }
+            valiElement(index);
+        }
+        function valiElement(index) {
+            $scope.addTherapyError = false;
+            var therapyVali = [];
+            _.each($scope.therapy, function(element) {
+                element = element.sort();
+                element = element.join(' ').trim();
+                if (element!=='')
+                    therapyVali.push(element);
+            });
+            for(var i=0;i<therapyVali.length;i++){
+                for(var j=i+1;j<=therapyVali.length;j++){
+                    if (therapyVali[i] == therapyVali[j]) {
+                        $scope.therapyErrorMessage = "Same Elements. Please check and save again.";
+                        $scope.addTherapyError = true;
+                    }
+                }
+            }
+        }
+
+        $scope.deleteTherapy = function(index){
+            if (index > 0){
+                $scope.therapy.splice(index, 1);
+            }
+            valiElement(index);
+        }
+
+        function isValidTreatment(indices, newTreatmentName) {
+            var isValid = true;
+            _.find(data.treatmentRef.treatments, function (treatment) {
+                if (treatment.name === newTreatmentName) {
+                    isValid = false;
+                    return true;
+                }
+            });
+            if (!isValid) {
+                $scope.therapyErrorMessage = "Same therapy exists.";
+                $scope.addTherapyError = true;
+            }
+            return isValid;
+        }
+
+        $scope.save = function(){
+            for(;$scope.therapy[$scope.therapy.length-1]=='';){
```

Comment:
What's this for loop used for?

Reply:
To make sure that there is no blank therapy. For example, the user clicks the second element, the third element will show. If she didn't input anything in the second and third elements and didn't delete them, this function would help delete them.